### PR TITLE
[SPARK-33890][SQL] Improve the implement of trim/trimleft/trimright

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The current implement of trim/trimleft/trimright have somewhat redundant.

### Why are the changes needed?
Improve the implement of trim/trimleft/trimright

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
Jenkins test

### CODE REVIEW

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
cc @cloud-fan  @wangyum

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -756,6 +756,52 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
...
+  }
+
+  protected val trimMethod: String
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val evals = children.map(_.genCode(ctx))
+    val srcString = evals(0)
+
+    if (evals.length == 1) {
+      ev.copy(evals.map(_.code) :+ code"""
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:777]: nit: let's use the
```
"""
  |xxx
  |xxx""".stripMargin
```
syntax

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -756,6 +756,52 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
...
+        if (${srcString.isNull}) {
+          ${ev.isNull} = true;
+        } else {
+          ${ev.value} = ${srcString.value}.$trimMethod();
+        }""")
+    } else {
+      val trimString = evals(1)
+      val getTrimFunction =
+        s"""
+        if (${trimString.isNull}) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:789]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -756,6 +756,52 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
...
+        boolean ${ev.isNull} = false;
+        UTF8String ${ev.value} = null;
+        if (${srcString.isNull}) {
+          ${ev.isNull} = true;
+        } else {
+          ${ev.value} = ${srcString.value}.$trimMethod();
+        }""")
+    } else {
+      val trimString = evals(1)
+      val getTrimFunction =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:787]: can we inline it and use `if... else if... else`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 4]
```diff
@@ -756,6 +756,52 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
   override def nullable: Boolean = children.exists(_.nullable)
   override def foldable: Boolean = children.forall(_.foldable)
 
+  protected def doEval(input: InternalRow, srcString: UTF8String): Any
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:759]: how about `doEval(srcString: UTF8String, trimString: Option[UTF8String])`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 1 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:777]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:787]: Good idea.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 5]
```diff
@@ -756,6 +756,54 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
   override def nullable: Boolean = children.exists(_.nullable)
   override def foldable: Boolean = children.forall(_.foldable)
 
+  protected def doEval(srcString: UTF8String): Any
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:759]: These will definitely return UTF8String right?

**Review by srowen — COMMENTED** [review]
Looks fine as a refactoring.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 5 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Yes. I updated it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 6]
```diff
@@ -756,6 +756,54 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
...
+  }
+
+  protected val trimMethod: String
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val evals = children.map(_.genCode(ctx))
+    val srcString = evals(0)
+
+    if (evals.length == 1) {
+      ev.copy(code = evals.map(_.code) :+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:780]: nit:
```
ev.copy(code = code"""
  |${evals.head.code}
  |...""".stripMargin 
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 7]
```diff
@@ -756,6 +756,54 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
...
+              |UTF8String ${ev.value} = null;
+              |if (${srcString.isNull}) {
+              |  ${ev.isNull} = true;
+              |} else {
+              |  ${ev.value} = ${srcString.value}.$trimMethod();
+              |}
+            """)
+    } else {
+      val trimString = evals(1)
+      ev.copy(code = evals.map(_.code) :+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:792]: We can skip evaluating trim string if possible
```
ev.copy(code = code"""
  |${evals.head.code}
  |if (${srcString.isNull}) {
  | ...
  |} else {
  |  ${trimString.code}
  |  if (${trimString.isNull}) ...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 6 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:780]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 7 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:792]: Good idea.

**Comment by beliefer** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thanks for your help! @srowen Thanks for your review.

**Pull request closed without merging**